### ISSUE TITLE
[Bug] Stale hostnames are used when creating new connections with IP addresses

### ISSUE THREAD

**Issue by lujjjh** [Dreamacro/clash#2847]
### Prerequisites

- [X] I understand that this is the official open-source version of the Clash core, **only providing support for the open-source version or Premium version**
- [X] I am submitting an issue with the Clash core, not Clash.Meta / OpenClash / ClashX / Clash For Windows or any other derivative version
- [X] I am using the latest version of the Clash or Clash Premium core **in this repository**
- [X] I have searched at the [Issue Tracker](……/) **and have not found any related issues**
- [X] I have read the [official Wiki](https://dreamacro.github.io/clash/) **and was unable to solve the issue**
- [ ] (required for Premium core) I've tried the `dev` branch and the issue still exists

### Version

Clash 2023.06.30 linux amd64 with go1.20.5 Fri Jun 30 01:08:28 PM UTC 2023

### Configuration File

```YAML
dns:
  default-nameserver:
  - 223.5.5.5
  - 223.6.6.6
  enable: true
  enhanced-mode: fake-ip
  fake-ip-filter:
  - foo.example.com
```

### Description

Let's say we connect to `foo.example.com` through Clash. Currently, it's pointed to 1.0.0.1 and it's on the `fake-ip-filter` list. This prevents it from being resolved to a fake IP address. The next time we try to connect to `1.0.0.1`, Clash will translate the IP address back to `foo.example.com` and establish a connection to that instead.

In case `foo.example.com` is updated to point to `127.0.0.1`, even if the TTL has expired, Clash will still connect to `foo.example.com` instead of directly connecting to `1.0.0.1`. This will result in connecting to `127.0.0.1` instead of `1.0.0.1`, even though the intention was to connect to the latter. And it appears that the only solution to workaround this issue is to restart Clash.

I've noticed that this behavior appears to be intentional:

https://github.com/Dreamacro/clash/blob/c7e34bdc1171cf749a658b4cd805de54fdc3d97d/dns/enhancer.go#L81

But I still believe this is a bug and should be addressed.

### Reproduction Steps

1. Add `foo.example.com` to the `fake-ip-filter` list.
2. Point `foo.example.com` to `1.0.0.1` and set the TTL to 1 minute.
3. When `curl foo.example.com` is executed, it receives a 301 response from Cloudflare.
4. When `curl 1.0.0.1` is executed, it also receives a 301 response from Cloudflare.
5. Change `foo.example.com` record to `127.0.0.1`.
6. Wait for minutes...
7. `curl foo.example.com` reports an expected error of "connection refused" or "empty reply from server" since 80/tcp is not being listened to on the local machine.
8. `curl 1.0.0.1` reports an unexpected error of "connection refused" or "empty reply from server".

**Comment by lujjjh** [comment]
One solution I can think of is to purge the *stale* cache when receiving subsequent DNS queries for `foo.example.com` (aka in the step 7 above).

**Comment by Dreamacro** [comment]
I think remove `cache.WithStale` is enough. reverse ip to domain mapping does not require stale.

**Comment by lujjjh** [comment]
It appears that removing `cache.WithStale` would be acceptable. However, in cases where the TTL is extremely short, such as 0, it might be necessary to set a minimum caching duration to prevent issues with reverse lookups.

**Comment by Dreamacro** [comment]
fixed on c1d027d.

Do you want to join the premium repository? Just don't leak the source code.
